[Chuah Jia Jie] iP

chuahjiajie opened the pull request:
### DukePro

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~FAST~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2324s2.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x]  Managing tasks
- [ ]  Managing deadlines (coming soon)
- [ ]  Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Nauseous-Spartan committed:
Made changes according to "Level 0" - Changed name, added a greeting and an exit.

Nauseous-Spartan committed:
Level-1 Added Echo

Nauseous-Spartan committed:
Level-1 Added echo

Nauseous-Spartan committed:
Level-2

Nauseous-Spartan committed:
Level-3

Nauseous-Spartan committed:
Level-4

Nauseous-Spartan committed:
Ai Testing

Nauseous-Spartan committed:
Level-5

Nauseous-Spartan committed:
Level-6

Nauseous-Spartan committed:
Merge commit '701d2513ebccc40a420b834d21faa6ac75ee4fce'

# Conflicts:
#	src/main/java/OedipusRex.java

Nauseous-Spartan committed:
level-7

Nauseous-Spartan committed:
level-8

Nauseous-Spartan committed:
A-MoreOOP

Nauseous-Spartan committed:
A-Packages

Nauseous-Spartan committed:
A-Gradle

Nauseous-Spartan committed:
A-JUnit

Nauseous-Spartan committed:
A-Jar

Nauseous-Spartan committed:
A-JavaDoc

Nauseous-Spartan committed:
A-CodingStandard

Nauseous-Spartan committed:
Level-9

Nauseous-Spartan committed:
Add features "Add" and "List"

Nauseous-Spartan committed:
Add feature "Mark as Done"

Nauseous-Spartan committed:
Add partial changes for "ToDo, Deadline, Events" feature

Nauseous-Spartan committed:
Add Level-4 features "ToDos, Events, Deadlines"

Nauseous-Spartan committed:
Add Automated Text UI Testing

Nauseous-Spartan committed:
Add Level-5 features, Exception Handling

Nauseous-Spartan committed:
Add Level-5 features, exception handling

Nauseous-Spartan committed:
Add exceptions for mark and unmark commands

Nauseous-Spartan committed:
Add Level-6 features, deletion

Nauseous-Spartan committed:
Add "SaveAndLoad" class to handle saving and loading for Level-7

Nauseous-Spartan committed:
Merge branch 'branch-Level-7'

Nauseous-Spartan committed:
Add minor changes for save and load

Nauseous-Spartan committed:
Add minor changes for save and load

yeoshuheng commented on src/main/java/SaveAndLoad.java:
```diff
@@ -0,0 +1,174 @@
+        }
+    }
+
+    public static void load(Task[] listOfTasks) {
```
Maybe it be easier to save your logged values in a comma / hyphen separated format, the logged values could be the components needed for the constructor of your task. 

This would allow you to directly create the needed task using it's own constructor.

Additionally, this could also possibly make parsing the input easier too, as you can just call a .split(",") on a new line read by the buffered reader.

This could make this portion of the code lot clearer and reduce the code you have to manage :)

yeoshuheng replied:
Maybe it be easier to save your logged values in a comma / hyphen separated format, the logged values could be the components needed for the constructor of your task.

This would allow you to directly create the needed task using it's own constructor.

Additionally, this could also possibly make parsing the input easier too, as you can just call a .split(",") on a new line read by the buffered reader.

This could make this portion of the code lot clearer and reduce the code you have to manage :)

yeoshuheng commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,33 @@
+        this.isDone = false;
+    }
+
+    public String cleanWhiteSpace(String word) {
```
Maybe it could be easier to using java's .trim() method? Nonetheless, great work in trying to clean up the whitespaces!

yeoshuheng replied:
```suggestion
    public String cleanWhiteSpace(String word) {
        return word.trim();
    }
```

yeoshuheng commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,27 @@
+    protected String to;
+
+
+    public Event(String description, String from, String to) throws InvalidInputException{
```
Good handling of various types of input errors :)

yeoshuheng commented on src/main/java/BalkanBot.java:
```diff
@@ -0,0 +1,202 @@
+
+    public static void main(String[] args) {
+        String line = "------------------------------------------";
+        Task[] listOfTasks = new Task[100];
```
Rather than using a java array, perhaps it would be easier and more scalable to use a ArrayList instead?
```suggestion
        ArrayList<Task> listOfTasks = new ArrayList<>();
```

yeoshuheng commented on src/main/java/SaveAndLoad.java:
```diff
@@ -0,0 +1,174 @@
+
+public class SaveAndLoad {
+
+    enum TaskType {
```
Great use of enums :)

AwJoey replied:
I second this! 💯

yeoshuheng reviewed (commented):
Overall great job!

I liked that you tried to identify and handle a large amount of errors and your code is clean and easy to read with aptly named variables.

However, it would be great to see more your classes being organised in packages and abstracted out from each other as per the more OOP class.

Nauseous-Spartan committed:
Finish Date and Time feature

Nauseous-Spartan committed:
Add Storage, TaskList and Ui classes. Incomplete work in progess

AwJoey commented on src/main/java/BalkanBot.java:
```diff
@@ -0,0 +1,205 @@
+
+
+
+        while (true) {
```
Looks good! Perhaps you could move this command checks into a java class (e.g. Parser)?

AwJoey commented on src/main/java/BalkanBot.java:
```diff
@@ -0,0 +1,205 @@
+                        break;
+                    }
+                    case "unmark": {
+                        if (brokenCommand.length < 2) {
```
While reviewing other PRs, I have seen a couple of people splitting each command into classes, perhaps you could try that too! Alternatively, you could also try using an enum class (I've implemented that, you could have a look) 😄

AwJoey commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,43 @@
+
+    @Override
+    public String toString() {
+        return "[D]" + super.toString() + " (by: " + deadline.getMonth() + " " + deadline.getDayOfMonth() + " " +
```
Perhaps you could create a static variable for the tasktype abbreviations (i.e. "D")? This might make it easier to edit in the future if this abbreviation changes

AwJoey reviewed (commented):
Great work so far! 💯 Suggested some small tweaks :)

Happy lunar new year!!

Nauseous-Spartan committed:
Add files: Storage, TaskList, Ui, Parser and BalkanBotException

Nauseous-Spartan committed:
Change ui to not utilise scanner

Nauseous-Spartan committed:
Finish A-MoreOOP. Balkanbot.java, Parser.java, Storage.java fully completed

Nauseous-Spartan committed:
Merge branch 'A-MoreOOP-Branch'

Nauseous-Spartan committed:
Separate into packages

Nauseous-Spartan committed:
Separate into packages

Nauseous-Spartan committed:
Merge branch 'add-gradle-support' of https://github.com/chuahjiajie/ip

Nauseous-Spartan committed:
Add Level-10

Nauseous-Spartan committed:
Add JUnit Tests

Nauseous-Spartan committed:
Create JAR file

Nauseous-Spartan committed:
Add JavaDoc to the following classes: BalkanBot, Parser, Storage, Deadline, Event, Task, TaskList, ToDo and Ui. This is to provide more clarity on what each class and method does

Nauseous-Spartan committed:
Fix code to be up to coding standards to follow course requirement and to make it more readable for other programmers

Nauseous-Spartan committed:
Add "find" command to balkan bot which allows user to search for tasks with descriptions that contain a specific keyword

Nauseous-Spartan committed:
Merge branch 'branch-A-CodingStandard'

Nauseous-Spartan committed:
Merge branch 'branch-A-Level-9'

Nauseous-Spartan committed:
Changed variable names of a few lines of code to fix issues aroused during merge process

Nauseous-Spartan committed:
Make standard GUI based on tutorial provided by Week 4 GUI tutorial so that bot has GUI

Nauseous-Spartan committed:
Add fully functional GUI that contains images and integrates the previous functionality of the chatbot into the form of a GUI

Nauseous-Spartan committed:
Merge branch 'branch-A-Level-10'

* branch-A-Level-10:
  Add fully functional GUI that contains images and integrates the previous functionality of the chatbot into the form of a GUI
  Make standard GUI based on tutorial provided by Week 4 GUI tutorial so that bot has GUI

Nauseous-Spartan committed:
Fix files for jar file so that eventually will have a working jar file that can be run on any computer

Nauseous-Spartan committed:
Fix issue where dialog box would cut off with elipsis when it was too long, dialog box is no longer cut off

Nauseous-Spartan committed:
Made changes to GUI

There was an issue with the GUI where the output box would not load
more than 5 lines worth of tasks when prompted to list out tasks.

Fixing this would allow for a better user experience.

As a step towards improving the GUI, let's make sure the code
in the MainWindow.fxml file is properly calibrated. Doing so will
make the subsequent implementation of GUIs functional.

Nauseous-Spartan committed:
Add Assertion to Classes

There were multiple areas in the Parser and other classes that
could have benefitted from the use of assertions.

Nauseous-Spartan committed:
Add Assertion Branch

Was going to implement assertions on a
separate branch, but I forgot.

Nauseous-Spartan committed:
Merge branch 'branch-A-Assertions'

* branch-A-Assertions:
  Add Assertion Branch

Nauseous-Spartan committed:
Add Level-8 Branch

Forgot to add level 8 branch lol.

Nauseous-Spartan committed:
Merge branch 'branch-level-8'

* branch-level-8:
  Add Level-8 Branch

Nauseous-Spartan committed:
Edit Coding Quality

Made adjustments to code quality

Nauseous-Spartan committed:
Fix Delete Errors

There was an error where the application would
display unexpected behaviour when a non positive
integer was provided for the delete command.

Nauseous-Spartan committed:
Correct Boolean Name

Changed boolean names in parser.

Nauseous-Spartan committed:
Merge branch 'branch-A-CodeQuality'

* branch-A-CodeQuality:
  Correct Boolean Name
  Fix Delete Errors
  Edit Coding Quality

Nauseous-Spartan committed:
Add Help feature

Chatbot now has a help command.
User simply needs to type in "help" and is able to
see a full list of all the commands for the bot.

Nauseous-Spartan committed:
Merge branch 'branch-C-Help'

* branch-C-Help:
  Add Help feature

Nauseous-Spartan committed:
Fix Date Time Feature

Errors found and corrected in deadline and event
classes due to issues with date time feature.

Nauseous-Spartan committed:
Add Ui.png

Added picture for Ui.png into
docs folder.

Nauseous-Spartan committed:
Fix test cases

Previous test cases no longer valid.
Updated with new test cases.

Nauseous-Spartan committed:
Add Feature 1

Feature 1 was not added.
Adding for fun.

chuahjiajie committed:
Merge pull request #2 from chuahjiajie/feature1

Add Feature 1

Nauseous-Spartan committed:
Finalise Bot v0.2

Final commit for project.

Nauseous-Spartan committed:
Add User Guide

Added a user guide for Balkan Bot.

Nauseous-Spartan committed:
Change User Guide

Made changes to user guide.

chuahjiajie committed:
Fix Storage

Storage file had methods that were
too long.

chuahjiajie committed:
Fix Parser

Parser Method was too long.

chuahjiajie committed:
Fix Minor Formatting
